File: Week 4/Thirteen.py
import sys, re, operator, string

# Auxiliary functions that can't be lambdas
def extract_words(obj, path_to_file):
    print(f"Trying to open text file: {path_to_file}")  # Debugging print
    try:
        with open(path_to_file) as f:
            obj['data'] = f.read()
    except FileNotFoundError:
        print(f"Error: File '{path_to_file}' not found.")
        sys.exit(1)
    
    pattern = re.compile(r'[\W_]+')
    data_str = ''.join(pattern.sub(' ', obj['data']).lower())
    obj['data'] = data_str.split()

def load_stop_words(obj):
    stop_words_path = '../stop_words.txt'
    print(f"Trying to open stop words file: {stop_words_path}")  # Debugging print
    try:
        with open(stop_words_path) as f:
            obj['stop_words'] = f.read().split(',')
    except FileNotFoundError:
        print(f"Error: Stop words file '{stop_words_path}' not found.")
        sys.exit(1)
    
    obj['stop_words'].extend(list(string.ascii_lowercase))

def increment_count(obj, w):
    obj['freqs'][w] = 1 if w not in obj['freqs'] else obj['freqs'][w] + 1

def create_object(initial_data):
    this = initial_data

    def get_attr(name):
        return this[name]

    def set_attr(name, value):
        this[name] = value

    return {'get': get_attr, 'set': set_attr, **this}

data_storage_obj = create_object({
    'data': [],
    'init': lambda path_to_file: extract_words(data_storage_obj, path_to_file),
    'words': lambda: data_storage_obj['get']('data')
})

stop_words_obj = create_object({
    'stop_words': [],
    'init': lambda: load_stop_words(stop_words_obj),
    'is_stop_word': lambda word: word in stop_words_obj['get']('stop_words')
})

word_freqs_obj = create_object({
    'freqs': {},
    'increment_count': lambda w: increment_count(word_freqs_obj, w),
    'sorted': lambda: sorted(word_freqs_obj['get']('freqs').items(), key=operator.itemgetter(1), reverse=True),
    'top25': lambda: [
        print(w, '-', c) for (w, c) in word_freqs_obj['get']('sorted')()[0:25]
    ]
})

# Initialize data and stop words
data_storage_obj['init'](sys.argv[1])
print(f"Loaded {len(data_storage_obj['words']())} words.")
stop_words_obj['init']()
print(f"Loaded {len(stop_words_obj['get']('stop_words'))} stop words.")

# Count word frequencies, ignoring stop words
for w in data_storage_obj['words']():
    if not stop_words_obj['is_stop_word'](w):
        word_freqs_obj['increment_count'](w)
        print(f"Counting: {w}")

# Call the top25 method to print the top 25 word frequencies
word_freqs_obj['top25']()
print("Top 25 printed.")

File: Week 4/Thirteen.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::io::Read;

// Utility functions
fn extract_words(data_storage: &mut HashMap<&str, String>, path_to_file: &str) {
    let mut file_content = String::new();
    if let Ok(mut file) = fs::File::open(path_to_file) {
        file.read_to_string(&mut file_content).unwrap();
    }
    let cleaned_content = file_content
        .to_lowercase()
        .replace(|c: char| !c.is_alphanumeric() && !c.is_whitespace(), " ");
    let words = cleaned_content.split_whitespace().map(String::from).collect::<Vec<_>>().join(" ");
    data_storage.insert("data", words);
}

fn load_stop_words(stop_words: &mut HashMap<&str, Vec<String>>) {
    if let Ok(content) = fs::read_to_string("../stop_words.txt") {
        let mut stop_words_vec: Vec<String> = content.split(',').map(String::from).collect();
        stop_words_vec.extend(('a'..='z').map(|c| c.to_string()));
        stop_words.insert("stop_words", stop_words_vec);
    }
}

fn increment_count(word_freqs: &mut HashMap<String, u32>, word: &str) {
    let count = word_freqs.entry(word.to_string()).or_insert(0);
    *count += 1;
}

fn is_stop_word(stop_words: &HashMap<&str, Vec<String>>, word: &str) -> bool {
    stop_words
        .get("stop_words")
        .map_or(false, |sw| sw.contains(&word.to_string()))
}

fn sorted(word_freqs: &HashMap<String, u32>) -> Vec<(String, u32)> {
    let mut freqs_vec: Vec<(String, u32)> = word_freqs.iter().map(|(k, &v)| (k.clone(), v)).collect();
    freqs_vec.sort_by(|a, b| b.1.cmp(&a.1));
    freqs_vec
}

// Main 'things'
fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Please provide a file name.");
        return;
    }
    let file_path = &args[1];

    // Data storage object equivalent
    let mut data_storage_obj: HashMap<&str, String> = HashMap::new();
    extract_words(&mut data_storage_obj, file_path);

    // Stop words object equivalent
    let mut stop_words_obj: HashMap<&str, Vec<String>> = HashMap::new();
    load_stop_words(&mut stop_words_obj);

    // Word frequencies object equivalent
    let mut word_freqs_obj: HashMap<String, u32> = HashMap::new();

    // Processing
    if let Some(data) = data_storage_obj.get("data") {
        let words: Vec<&str> = data.split_whitespace().collect();
        for word in words {
            if !is_stop_word(&stop_words_obj, word) {
                increment_count(&mut word_freqs_obj, word);
            }
        }
    }

    // Sorting and printing
    let sorted_freqs = sorted(&word_freqs_obj);
    for (word, count) in sorted_freqs.iter().take(25) {
        println!("{} - {}", word, count);
    }
}